impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -1355,9 +1373,16 @@ private void expireMapping(final K key, final ValueHolder<V> value) {
       }
     }
 
-    return makeValue(key, newValue, now, expirationTime, this.valueCopier);
+    OnHeapValueHolder<V> holder = null;
+    try {
+      holder = makeValue(key, newValue, now, expirationTime, this.valueCopier);
+      eventSink.updated(key, oldValue, newValue);
+    } catch (LimitExceededException e) {
+      LOG.warn(e.getMessage());
```

Comment:
There is a missing removed event here. See what happens when expiry result is a duration of zero.